### ISSUE TITLE
Status of OPUS Codec

### ISSUE THREAD

**Issue by perfectmak** [pdeljanov/Sonata#8]
What is the status of work on the opus Codec? I see it marked as `next` in the readme, is there an ETA for release?

**Comment by pdeljanov** [comment]
Work on Opus hasn't started yet, but it's the next item on the road map.

Before I can begin work on Opus the OGG branch must be merged. However, before I can do that I first need to re-work some of Sonata's public API. This work is about 80% complete.

**Comment by perfectmak** [comment]
Okay. Thanks for the update. I'll keep an eye on the repository in the meantime.

**Comment by Gelbpunkt** [comment]
Hi, any status updates on this? I'd love to replace libopus with a pure rust thing in one of my projects and Symphonia looks most promising.

**Comment by pdeljanov** [comment]
Hi @Gelbpunkt,

The prerequisite work noted above has been completed, however I'm afraid to say that not much progress has been made on the Opus decoder itself.

I'm on a bit of a temporary hiatus from my personal projects at the moment so I wouldn't expect anything soon.

**Comment by Beiri22** [comment]
Some time has passed; any news on opus?

**Comment by pdeljanov** [comment]
Sorry to say, but there's not much to report. :confused:  Haven't had the time to do much beyond bug fixes. I'd suggest looking for alternative crates if you need Opus support.

**Comment by thinking-tower** [comment]
Hey @pdeljanov, can I give this a go?

**Comment by pdeljanov** [comment]
@thinking-tower, please, go ahead. You can use the `opus` branch to get you started. It contains all the required boilerplate and gets you to a point where you can work on decoding the audio packets. It just needs to be brought up-to date first. Best of luck!

**Comment by thinking-tower** [comment]
@pdeljanov I'm still working on this ticket btw! Thanks for your patience 😄

**Comment by pdeljanov** [comment]
Hey @thinking-tower, that's good to hear! Are you running into any challenges or issues?

Opus is a pretty complex codec, so feel free to submit multiple PRs to the `opus` branch. I believe that Opus will prompt some changes in `symphonia-core`, for example, adding arithmetic decoding to the `BitReader`s, or more DSP routines to the `dsp` module. We could collaborate on these changes if there is a central branch we can work off of.

**Comment by thinking-tower** [comment]
@pdeljanov At the moment, not any issues yet. I like the idea of splitting up the commits into multiple PRs!

**Comment by darksv** [comment]
Hi @thinking-tower! Are you actively working on a decoder for Opus? I would like to help with the implementation.

**Comment by pdeljanov** [comment]
Hi @darksv,

Thanks for volunteering to work on this. Let's wait a bit longer to give a chance for @thinking-tower to respond, but as far as I know the Opus decoder hasn't been actively worked on in a long time.

There are two Opus related branches. The first is the `opus` branch which was my first attempt from two years ago before I focused my attention onto other priorities. The second, newer, branch is `opus-dev` which is where @thinking-tower was submitting PRs to.

It would probably be worthwhile to first bring `opus-dev` up-to-date, then port the remaining code from `opus` over to it. We can then make `opus-dev` the official Opus branch going forward.

I think for Opus you should aim for smaller PRs that can be merged into `opus-dev`. That way I, and other developers, can contribute to the effort. Opus is really two codecs in one, and will also likely drive many changes to `symphonia-core` and `symphonia-utils-xiph`.

**Comment by thinking-tower** [comment]
Hi @darksv and @pdeljanov, I am still working on the Opus decoder. I'm definitely happy to receive help!

I'm currently just debugging the official repo (https://github.com/xiph/opus) to get a better feel of how the decoding process works, are you familiar with Opus yourself?

@pdeljanov Thanks again for your patience!

**Comment by darksv** [comment]
> are you familiar with Opus yourself?

@thinking-tower Not really. I've read some parts of the RFC and looked into the official implementation a bit, but I'm not very familiar with Opus just yet.

**Comment by SunnyWar** [comment]
Perhaps this work should be divided into SILK and CELT sub-codecs. SILK being used for lower bandwidth speech and CELT is for higher bandwidth music, typically.

**Comment by echelon** [comment]
I'll pay a bounty (~$500) to whomever lands `opus` / `e1a70095c4409478c1ded499fb8dd4afa540fb90`. 

This is important for dealing with web uploads as browser APIs use opus. 

Bonus points if media duration is easy to extract.

**Comment by chris-ha458** [comment]
I will match  @echelon's offer.
Also 100$ partial bounties for : 
1. bringing opus-dev recent upto master
2. any complete SILK implementation
3. any complete CELT implementation
4. if 1 and 2 has been achieved integration of 1,2 algorithms
5. any sub task as @pdeljanov sees fit.

all determinations as to (sub-)task completion decision is upto @pdeljanov and obviously these are payable to @pdeljanov as well (in this case I will request for tests to be implemented as well for objective review of achievement)

**Comment by chris-ha458** [comment]
I think i might start with 1 myself

**Comment by JSDurand** [comment]
I just found this repository [opus-native](https://github.com/hasenbanck/opus-native), which aims to implement opus codec in pure Rust.  
From a rough glance at its codes, it seems that it can help others implement the codec at least.  
Maybe it is worth sharing this information here.  :)

**Comment by chris-ha458** [comment]
@JSDurand thanks for mentioning [opus-native](https://github.com/hasenbanck/opus-native).
But it seems that not only do the tests do not pass, they don't even compile atm.

**Comment by JSDurand** [comment]
@chris-ha458 Yes, that is not finished.  But it did implement some decoding logics, such as the range decoding, and the outer-most decoder logic.  I am not familiar with this topic, so I cannot tell how completed it is, though.  Just thought this might help the implementers if they did not know about that repository previously.

**Comment by thinking-tower** [comment]
@JSDurand thanks for that link, it'll be helpful for me.

@chris-ha458 How are you getting on with this? I've got some time to work on this now.

**Comment by chris-ha458** [comment]
I think the only fair and reasonable approach would be to defer to the maintainer atm.

**Comment by thinking-tower** [comment]
I was working on this before, so I'll continue. I'm happy to work with anyone interested.

**Comment by FreezyLemon** [comment]
Hey, can I ask what the current state of this is? I would assume the API changes mentioned earlier are done and that the OGG implementation is now at a point where the opus codec can be implemented?

This looks interesting and I might take a shot at it over the weekends (no promises though). @thinking-tower are you still actively working on this? And if so, what are you working on atm?

**Comment by giorgi-o** [comment]
Hi, any updates on this? @pdeljanov @thinking-tower

**Comment by Enitoni** [comment]
Bumping this, as I too would really like opus support.

**Comment by n1ght-hunter** [comment]
+1

**Comment by CosmicHorrorDev** [comment]
(stop spamming the thread with junk that 99% of people who are subscribed don't want to see)

**Comment by Enitoni** [comment]
> (stop spamming the thread with junk that 99% of people who are subscribed don't want to see)

Spam? The last message (from me) was 4 days ago. It's pretty common to bump issues that have high demand. You can press this button if you don't want to see it:
![bilde](https://github.com/pdeljanov/Symphonia/assets/5295397/15b46143-7c52-40c6-bc6a-81df19178cff)

**Comment by CosmicHorrorDev** [comment]
To be clear I was referring to the message from just a few minutes ago, not yours from 4 days ago

The problem is that I **do** want to be subscribed, but I don't want "+1" notifications when I'm subscribed to hundreds of issues

**Comment by Enitoni** [comment]
> To be clear I was referring to the message from just a few minutes ago, not yours from 4 days ago
> 
> The problem is that I **do** want to be subscribed, but I don't want "+1" notifications when I'm subscribed to hundreds of issues

That is fair, but I personally don't see any semantic difference between "+1" and "I too would like this this supported", so that's why I was confused.

**Comment by giorgi-o** [comment]
Can https://github.com/jiztastamablastamarang/symphonia/pull/1 be merged into this repo?

It looks like that person wanted to get certified in Rust, and chose to implement Opus (SILK) in Symphonia as the "final project" of the certification process. So, it appears to be in a separate fork only due to the [certification process requiring it to be that way](https://github.com/rust-lang-ua/rustcamp/tree/master/6_project).

**Comment by pdeljanov** [comment]
This was submitted as a PR some time ago, but it was promptly closed by the author before I could take a look at it. It looked interesting, but the authors need to be willing for me to merge it! Alternatively, if they don't want to, they could look into publishing it on crates.io directly. A user would just need to register it manually into Symphonia.

Just in general, my focus at this time is on completing work on the 0.6 version of Symphonia (see `dev-0.6` branch). After that, Opus is my top priority for this project.

**Comment by aschey** [comment]
It's obviously not as good as a first-party solution, but I've just released [`symphonia-adapter-libopus`](https://github.com/aschey/symphonia-adapters/tree/main/crates/symphonia-adapter-libopus), which wraps the [libopus Rust bindings](https://github.com/DoumanAsh/opusic-sys) in a Symphonia `Decoder`. Seems to integrate well with the rest of Symphonia in my testing so far.
